File: 1123_lca_depth.py
def solve(root):
    store = []
    max_depth = 0

    if not root:
        return None
    
    print(helper(root, 0))


def helper(root, depth):
    if not root:
        return None, depth

    leftLCA, leftDepth = helper(root.left, depth+1)
    rightLCA, rightDepth = helper(root.right, depth+1)

    if leftDepth == rightDepth:
        return leftDepth, root.val
    elif leftDepth > rightDepth:
        return leftDepth, leftLCA
    else:
        return rightDepth, rightLCA

File: 1123_lca_depth.cpp
TreeNode* helper(TreeNode* root, int& depth) {
  if (root == nullptr) return nullptr;

  int leftDepth = depth + 1, rightDepth = depth + 1;
  TreeNode* leftLCA = helper(root->left, leftDepth);
  TreeNode* rightLCA = helper(root->right, rightDepth);

  if (leftDepth == rightDepth) {
    depth = leftDepth;
    return root;
  } else if (leftDepth > rightDepth) {
    depth = leftDepth;
    return leftLCA;
  } else {
    depth = rightDepth;
    return rightLCA;
  }
}
